File: backend/routes/advisor.py
from flask import Blueprint, request, jsonify, current_app
from flask_jwt_extended import jwt_required, get_jwt_identity
from database import db
from models import User, Budget, Goal, ChatLog
from google import genai
from google.genai import types
import os

advisor_bp = Blueprint('advisor', __name__)


def build_system_prompt(user, budgets, goals):
    latest    = budgets[0] if budgets else None
    income    = latest.income if latest else 0
    expenses  = latest.expenses if latest else []
    total_exp = sum(e.actual for e in expenses)
    savings   = income - total_exp
    rate      = round((savings / income) * 100, 1) if income else 0

    cat_lines = '\n'.join(
        f"  - {e.category}: Rs.{e.actual}"
        + (f" (budget: Rs.{e.budget_amt})" if e.budget_amt else "")
        for e in expenses if e.actual > 0
    ) or '  No expenses recorded yet'

    goal_lines = '\n'.join(
        f"  - {g.name}: Rs.{g.saved} saved of Rs.{g.target} target"
        + (f" (by {g.target_date})" if g.target_date else "")
        for g in goals
    ) or '  No goals set yet'

    return f"""You are FinTech, a warm and expert Indian personal finance advisor.

USER PROFILE:
- Name: {user.name}
- City type: {user.city_type}

CURRENT FINANCES:
- Monthly Income: Rs.{income:,.0f}
- Total Monthly Expenses: Rs.{total_exp:,.0f}
- Monthly Savings: Rs.{savings:,.0f}
- Savings Rate: {rate}%

EXPENSE BREAKDOWN:
{cat_lines}

FINANCIAL GOALS:
{goal_lines}

INSTRUCTIONS:
- Personalise every response using the user's actual numbers above.
- Suggest Indian instruments: SIP, PPF, ELSS, NPS, FD, RD, NSC.
- Mention tax benefits under 80C and 80D where relevant.
- Use Indian number formatting (rupees, lakhs, crores).
- Be warm, practical and encouraging.
- Keep responses under 200 words unless asked for more detail.
- Use bullet points for lists and bold for key numbers."""


# ── CHAT ──────────────────────────────────────────────────────────
@advisor_bp.route('/chat', methods=['POST'])
@jwt_required()
def chat():
    user_id = int(get_jwt_identity())
    data    = request.get_json()
    message = data.get('message', '').strip()

    if not message:
        return jsonify({'error': 'Message cannot be empty'}), 400

    user    = User.query.get_or_404(user_id)
    budgets = Budget.query.filter_by(user_id=user_id).order_by(Budget.year.desc(), Budget.id.desc()).limit(3).all()
    goals   = Goal.query.filter_by(user_id=user_id).all()

    recent_logs = (ChatLog.query
                   .filter_by(user_id=user_id)
                   .order_by(ChatLog.created_at.asc())
                   .limit(10).all())

    # Save user message
    db.session.add(ChatLog(user_id=user_id, role='user', message=message))
    db.session.flush()

    try:
        api_key = os.getenv('GEMINI_API_KEY') or current_app.config.get('GEMINI_API_KEY', '')
        if not api_key:
            return jsonify({'error': 'Gemini API key not configured. Add GEMINI_API_KEY to your .env file'}), 500

        client = genai.Client(api_key=api_key)

        # Build conversation history
        history = []
        for log in recent_logs:
            role = 'user' if log.role == 'user' else 'model'
            history.append(types.Content(role=role, parts=[types.Part(text=log.message)]))

        # Add current message
        history.append(types.Content(role='user', parts=[types.Part(text=message)]))

        response = client.models.generate_content(
            model='gemini-2.0-flash',
            config=types.GenerateContentConfig(
                system_instruction=build_system_prompt(user, budgets, goals),
                max_output_tokens=1000,
                temperature=0.7,
            ),
            contents=history,
        )

        reply = response.text

        # Save assistant reply
        db.session.add(ChatLog(user_id=user_id, role='assistant', message=reply))
        db.session.commit()

        return jsonify({'reply': reply, 'role': 'assistant'}), 200

    except Exception as e:
        db.session.rollback()
        return jsonify({'error': f'AI service error: {str(e)}'}), 500


# ── GET CHAT HISTORY ──────────────────────────────────────────────
@advisor_bp.route('/history', methods=['GET'])
@jwt_required()
def get_history():
    user_id = int(get_jwt_identity())
    logs    = (ChatLog.query
               .filter_by(user_id=user_id)
               .order_by(ChatLog.created_at.asc())
               .limit(50).all())
    return jsonify({'history': [log.to_dict() for log in logs]}), 200


# ── CLEAR CHAT HISTORY ────────────────────────────────────────────
@advisor_bp.route('/history', methods=['DELETE'])
@jwt_required()
def clear_history():
    user_id = int(get_jwt_identity())
    ChatLog.query.filter_by(user_id=user_id).delete()
    db.session.commit()
    return jsonify({'message': 'Chat history cleared'}), 200

File: js/advisor.js
// ===== AI ADVISOR =====
const QUICK_QUESTIONS = [
  'How should I invest my monthly savings?',
  'Am I spending too much on rent?',
  'How to build a 6-month emergency fund?',
  'Best SIP funds for beginners in India?',
  'How can I reduce my monthly expenses?',
  'Explain PPF vs ELSS for tax saving',
  'What is the 50-30-20 rule?',
  'How to plan for retirement at 60?',
];

function renderQuickQs() {
  const el = document.getElementById('quick-qs');
  el.innerHTML = QUICK_QUESTIONS.map(q =>
    `<button class="quick-q-btn" onclick="askQuestion('${q.replace(/'/g,"\\'")}')">
      ${q}
    </button>`
  ).join('');
}

function askQuestion(q) {
  document.getElementById('chat-input').value = q;
  sendChat();
}

function buildSystemPrompt() {
  const totalExp = getTotalExpenses();
  const savings  = getSavings();
  const rate     = getSavingsRate();
  const score    = getHealthScore();

  const catBreakdown = state.categories
    .filter(c => c.actual > 0)
    .map(c => `  - ${c.name}: ₹${c.actual}${c.budget ? ' (budget: ₹' + c.budget + ')' : ''}`)
    .join('\n');

  const goalsList = state.goals.length
    ? state.goals.map(g => `  - ${g.name}: ₹${g.saved} saved of ₹${g.target} target${g.date ? ' by ' + g.date : ''}`).join('\n')
    : '  None set';

  return `You are FinTech — a friendly, expert personal finance advisor for India.

USER'S FINANCIAL PROFILE:
- Name: ${state.userName || 'User'}
- City type: ${state.city}
- Monthly Income: ₹${state.income}
- Total Monthly Expenses: ₹${totalExp}
- Monthly Savings: ₹${savings}
- Savings Rate: ${rate}%
- Financial Health Score: ${score}/100

EXPENSE BREAKDOWN:
${catBreakdown || '  No expenses entered yet'}

FINANCIAL GOALS:
${goalsList}

INSTRUCTIONS:
- Give personalized advice based on the user's actual numbers above.
- Always reference Indian financial instruments: SIP, PPF, ELSS, NPS, FD, RD, NSC, Sukanya Samriddhi, etc.
- Mention tax benefits under 80C, 80D where relevant.
- Use Indian number formatting (lakhs, crores).
- Be warm, encouraging, and practical. Avoid jargon unless explaining it.
- Keep responses concise (under 200 words unless the user asks for detail).
- If the user hasn't entered data, gently encourage them to do so for personalized advice.
- Format key numbers in bold using markdown **like this**.
- Use bullet points for lists.`;
}

function appendMessage(role, html) {
  const chatEl = document.getElementById('chat-messages');
  const div = document.createElement('div');
  div.className = 'chat-msg ' + role;
  const avatarText = role === 'ai' ? '✦' : (state.userName ? state.userName[0].toUpperCase() : 'U');
  const avatarClass = role === 'ai' ? 'ai-avatar' : 'user-avatar';
  div.innerHTML = `
    <div class="chat-avatar ${avatarClass}">${avatarText}</div>
    <div class="chat-bubble">${html}</div>`;
  chatEl.appendChild(div);
  chatEl.scrollTop = chatEl.scrollHeight;
  return div;
}

function simpleMarkdown(text) {
  return text
    .replace(/\*\*(.*?)\*\*/g, '<strong>$1</strong>')
    .replace(/\*(.*?)\*/g, '<em>$1</em>')
    .replace(/^### (.+)$/gm, '<h4 style="margin:8px 0 4px;font-size:14px;font-weight:600">$1</h4>')
    .replace(/^## (.+)$/gm, '<h3 style="margin:10px 0 6px;font-size:15px;font-weight:700">$1</h3>')
    .replace(/^- (.+)$/gm, '<li>$1</li>')
    .replace(/(<li>.*<\/li>)+/gs, '<ul style="margin:6px 0;padding-left:1.2rem">$&</ul>')
    .replace(/\n{2,}/g, '</p><p style="margin-top:8px">')
    .replace(/\n/g, '<br>');
}

let chatHistory = [];

async function sendChat() {
  const input = document.getElementById('chat-input');
  const msg   = input.value.trim();
  if (!msg) return;
  input.value = '';

  // User bubble
  appendMessage('user', msg);
  chatHistory.push({ role: 'user', content: msg });

  // Typing indicator
  const typingDiv = appendMessage('ai', '<div class="typing-dots"><span></span><span></span><span></span></div>');

  try {
    const response = await fetch('https://api.anthropic.com/v1/messages', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model: 'claude-sonnet-4-20250514',
        max_tokens: 1000,
        system: buildSystemPrompt(),
        messages: chatHistory,
      }),
    });

    const data  = await response.json();
    const reply = data.content?.map(b => b.text || '').join('') || 'I could not get a response. Please try again.';

    // Replace typing with actual response
    typingDiv.querySelector('.chat-bubble').innerHTML = simpleMarkdown(reply);
    chatHistory.push({ role: 'assistant', content: reply });

    // Keep history manageable
    if (chatHistory.length > 20) {
      chatHistory = chatHistory.slice(-18);
    }

  } catch (err) {
    typingDiv.querySelector('.chat-bubble').textContent =
      'Connection error. Please check your internet and try again.';
  }

  document.getElementById('chat-messages').scrollTop = 99999;
}